src/Lumi/Components/Form.purs

```diff
@@ -129,37 +138,175 @@ build editor = makeStateless (createComponent "Form") render where
              }
         contractProps = unsafeCoerce
 
-        fieldDivider = R.hr { className: "lumi field-divider" }
+     in defaultRenderForm
+          { readonly: props.readonly
+          , inlineTable: props.inlineTable
+          , forceTopLabels: props.forceTopLabels
+          }
+          forest
 
-        toRow = case _ of
-          Child { key, child } ->
-            maybe identity keyed key $ child
-          Wrapper { key, children } ->
-            R.div
-              { key: fromMaybe "" key
-              , children: [ intercalate fieldDivider (map toRow children) ]
-              }
-          Node { label, key, required, validationError, children } ->
-            maybe identity keyed key $ labeledField
-              { label: text body
-                  { children = [ label ]
-                  , className = toNullable (pure "field-label")
-                  }
-              , value: intercalate fieldDivider (map toRow children)
-              , validationError: validationError
-              , required: required
-              , forceTopLabel: forceTopLabels
-              , style: css {}
-              }
+-- | Render a form with state managed automatically.
+useForm
+  :: forall props unvalidated result xs
+   . RowToList unvalidated xs
+  => HMap ModifyValidated { | unvalidated } { | unvalidated }
+  => FormBuilder
+       { initialState :: { | unvalidated }
+       , readonly :: Boolean
+       , inlineTable :: Boolean
+       , forceTopLabels :: Boolean
+       | props
+       }
+       { | unvalidated }
+       result
+  -> { initialState :: { | unvalidated }
+     , readonly :: Boolean
+     , inlineTable :: Boolean
+     , forceTopLabels :: Boolean
+     | props
+     }
+  -> Hooks.Hook (Hooks.UseState { | unvalidated })
+      { formData :: { | unvalidated }
+      , setFormData :: ({ | unvalidated } -> { | unvalidated }) -> Effect Unit
+      , setModified :: Effect Unit
+      , reset :: Effect Unit
```

Comment:
Are these two (`setModified` and `reset`) necessary? My previous experience with providing these convenience values (as the final result in `onChange` was) wasn't the best :p

Reply:
`setModified` is just `setFormData F.setModified`, and `reset` is just `setFormData \_ -> initialState`.

Reply:
Nope, just convenience. I think you're right though, I don't want to reintroduce the async bugs. Some of these should probably be functions (in which case the convenience things are more useful too).